src/bun.js/api/bun/h2_frame_parser.zig

```diff
@@ -0,0 +1,2563 @@
+                    return JSC.JSValue.jsNumber(stream.id);
+                }
+                stream.weight = @intCast(weight);
+            }
+
+            if (options.get(globalObject, "signal")) |signal_arg| {
+                if (signal_arg.as(JSC.WebCore.AbortSignal)) |signal_| {
+                    if (signal_.aborted()) {
+                        stream.state = .CLOSED;
+                        stream.rstCode = @intFromEnum(ErrorCode.CANCEL);
+                        this.dispatchWithExtra(.onAborted, JSC.JSValue.jsNumber(stream.id), signal_.abortReason());
+                        return JSC.JSValue.jsNumber(stream.id);
+                    }
+                    stream.attachSignal(signal_);
+                }
+            }
+        }
+
+        var length: usize = encoded_size;
+        if (has_priority) {
+            length += 5;
+            flags |= @intFromEnum(HeadersFrameFlags.PRIORITY);
+        }
+
+        log("request encoded_size {}", .{encoded_size});
+        var frame: FrameHeader = .{
+            .type = @intFromEnum(FrameType.HTTP_FRAME_HEADERS),
+            .flags = flags,
+            .streamIdentifier = stream.id,
+            .length = @intCast(encoded_size),
+        };
+        var writer = this.toWriter();
+        frame.write(@TypeOf(writer), writer);
+        //https://datatracker.ietf.org/doc/html/rfc7540#section-6.2
+        if (has_priority) {
+            var stream_identifier: UInt31WithReserved = .{
+                .reserved = exclusive,
+                .uint31 = @intCast(parent),
+            };
+
+            var priority: StreamPriority = .{
+                .streamIdentifier = stream_identifier.toUInt32(),
+                .weight = @intCast(weight),
+            };
+
+            priority.write(@TypeOf(writer), writer);
+        }
+
+        this.write(buffer[0..encoded_size]);
+
+        if (end_stream) {
+            stream.state = .HALF_CLOSED_LOCAL;
+
+            if (waitForTrailers) {
+                this.dispatch(.onWantTrailers, JSC.JSValue.jsNumber(stream.id));
+            }
+        } else {
+            stream.waitForTrailers = waitForTrailers;
+        }
+
+        return JSC.JSValue.jsNumber(stream.id);
+    }
+
+    pub fn read(this: *H2FrameParser, globalObject: *JSC.JSGlobalObject, callframe: *JSC.CallFrame) callconv(.C) JSValue {
+        JSC.markBinding(@src());
+        const args_list = callframe.arguments(1);
+        if (args_list.len < 1) {
+            globalObject.throw("Expected 1 argument", .{});
+            return .zero;
+        }
+        const buffer = args_list.ptr[0];
+        if (buffer.asArrayBuffer(globalObject)) |array_buffer| {
+            var bytes = array_buffer.byteSlice();
+
+            // read all the bytes
+            while (bytes.len > 0) {
+                const result = this.readBytes(bytes);
+                if (result >= bytes.len) {
```

Comment:
Why are we checking if it read more than the amount? isn't the outer `while` loop enough here?

Reply:
fixed in https://github.com/oven-sh/bun/pull/6761/commits/946a993adea0d7084e96012fd60337da5be5bc49